# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [87]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import empiricaldist as ed
from scipy.stats import norm
from math import sqrt

%matplotlib inline
os.getcwd()
basepath = os.getcwd()
print(basepath)
lis_dir = os.listdir()
print(lis_dir)


C:\Users\Tiki\Documents\Sprinboard2\Guided_Capstone
['.ipynb_checkpoints', '1582673871_updated_ski_data.csv.zip', 'data', 'figures', 'GuidedCapstone_Steps4and5.ipynb', 'models', 'Notebook_stepfourandfive.ipynb', 'Notebook_stepthree.ipynb', 'Notebook_steptwo.ipynb', 'Notebook_steptwov2.ipynb', 'Ski_Lift_Problem_Statementsheet.pptx', 'step2_output.csv', 'step3_output.csv', 'updated_ski_data.csv', '__MACOSX']


**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [88]:
# load csv data
file ='step3_output.csv'
sub_file = 'data'
file_name_path = os.path.join(basepath, sub_file, file)
ski_data= pd.read_csv(file_name_path)
print(ski_data.dtypes)
ski_data.head(5)


Name                  object
state                 object
summit_elev            int64
vertical_drop        float64
trams                  int64
fastSixes            float64
fastQuads            float64
quad                 float64
triple               float64
double               float64
surface              float64
total_chairs         float64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object


Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090          294.0   
1               Arizona Snowbowl     Arizona        11500         2300.0   
2            Sunrise Park Resort     Arizona        11100         1800.0   
3  Yosemite Ski & Snowboard Area  California         7800          600.0   
4                    Bear Valley  California         8500         1900.0   

   trams  fastSixes  fastQuads  quad  triple  double  ...  SkiableTerrain_ac  \
0      0        0.0        0.0   0.0     1.0     0.0  ...               30.0   
1      0        1.0        0.0   2.0     2.0     1.0  ...              777.0   
2      0        0.0        1.0   2.0     3.0     1.0  ...              800.0   
3      0        0.0        0.0   0.0     1.0     3.0  ...               88.0   
4      0        1.0        1.0   0.0     2.0     4.0  ...             1680.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            30.0             150.0       36.0             69.0     30.000000   
1           104.0             122.0       81.0            260.0     89.000000   
2            80.0             115.0       49.0            250.0     74.000000   
3             0.0             110.0       84.0            300.0     47.000000   
4           100.0             165.0       52.0            359.0     57.916957   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0      34.00000              152.0            30.0         1  
1      89.00000              122.0             0.0         0  
2      78.00000              104.0            80.0         0  
3      47.00000              107.0             0.0         0  
4      64.16681              151.0             0.0         0  

[5 rows x 25 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [89]:
#dfo=df.select_dtypes(include=['object']) # select object type columns
#df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

# Create a set of dummy variables from the state variable
popped_state = ski_data['state']
ski_data = pd.concat([ski_data.drop(columns=['state']), pd.get_dummies(ski_data, columns=['state'])], axis=1)
ski_data.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090          294.0      0   
1               Arizona Snowbowl        11500         2300.0      0   
2            Sunrise Park Resort        11100         1800.0      0   
3  Yosemite Ski & Snowboard Area         7800          600.0      0   
4                    Bear Valley         8500         1900.0      0   

   fastSixes  fastQuads  quad  triple  double  surface  ...  \
0        0.0        0.0   0.0     1.0     0.0      2.0  ...   
1        1.0        0.0   2.0     2.0     1.0      2.0  ...   
2        0.0        1.0   2.0     3.0     1.0      0.0  ...   
3        0.0        0.0   0.0     1.0     3.0      1.0  ...   
4        1.0        1.0   0.0     2.0     4.0      2.0  ...   

   state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                   0                   0                0           0   
1                   0                   0                0           0   
2                   0                   0                0           0   
3                   0                   0                0           0   
4                   0                   0                0           0   

   state_Vermont  state_Virginia  state_Washington  state_West Virginia  \
0              0               0                 0                    0   
1              0               0                 0                    0   
2              0               0                 0                    0   
3              0               0                 0                    0   
4              0               0                 0                    0   

   state_Wisconsin  state_Wyoming  
0                0              0  
1                0              0  
2                0              0  
3                0              0  
4                0              0  

[5 rows x 83 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [90]:
#Was there a cell missing here for us to standardize data?

#drop name
name_popped = ski_data.pop('Name')

#import  sklearn
from sklearn import preprocessing

from sklearn.preprocessing import MinMaxScaler 
scaler = MinMaxScaler()


# separate the data and target attributes
ski_data_target = ski_data[['daysOpenLastYear', 'AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ]]
ski_data_numeric = ski_data.drop(['AdultWeekend'], axis =1)
ski_data_numeric = ski_data_numeric.loc[:, ~ski_data_numeric.columns.str.startswith('state')]
#ski_data_numeric2 = ski_data.drop(['daysOpenLastYear', 'AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ], axis =1)

print(ski_data.head())
#standardize numeric_ski_Data
list_col = ski_data_numeric.columns
ski_data[list_col] = scaler.fit_transform(ski_data[list_col])
print(ski_data.head())

   summit_elev  vertical_drop  trams  fastSixes  fastQuads  quad  triple  \
0         2090          294.0      0        0.0        0.0   0.0     1.0   
1        11500         2300.0      0        1.0        0.0   2.0     2.0   
2        11100         1800.0      0        0.0        1.0   2.0     3.0   
3         7800          600.0      0        0.0        0.0   0.0     1.0   
4         8500         1900.0      0        1.0        1.0   0.0     2.0   

   double  surface  total_chairs  ...  state_Rhode Island  state_South Dakota  \
0     0.0      2.0           3.0  ...                   0                   0   
1     1.0      2.0           8.0  ...                   0                   0   
2     1.0      0.0           7.0  ...                   0                   0   
3     3.0      1.0           5.0  ...                   0                   0   
4     4.0      2.0          10.0  ...                   0                   0   

   state_Tennessee  state_Utah  state_Vermont  state_Vir

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [91]:
from sklearn.model_selection import train_test_split
#ski_data_trial = ski_data

ski_data_trial = ski_data.drop(['AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ], axis =1)
# we will also be looking for projected days open so that can't be included
#The above difference was dropped due to the Rsq not being correct to model

y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial , y, test_size=0.25, random_state=1)




Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [92]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error, mean_squared_error
from sklearn.metrics import r2_score
lm = linear_model.LinearRegression()
lin_model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)

lin_model_coef = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef[col_name] = lm.coef_[0][idx]
    
coef_table =pd.DataFrame(lin_model_coef.items(), columns=['variable', 'coefficent'])


**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [93]:

lin_model_expl_var = explained_variance_score(y_test, y_pred)
lin_model_me = mean_absolute_error(y_test, y_pred)
lin_model_rsq = r2_score(y_test, y_pred)
print('The Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model_Rsq, lin_model_me))


The Rsq of the linear model = 0.3898894046548559 and the Mean Absolute Error = 623337519347.5455


## Review Model Outcomes — Iterate over additional models as needed.

In [94]:
from sklearn import metrics
from sklearn import linear_model
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import GridSearchCV, train_test_split

def test(models, X_train1, X_test1, y_train1, y_test1, iterations = 100):
    results = {}
    for i in models:
        r2_train = []
        r2_test = []
        for j in range(iterations):
            r2_test.append(metrics.r2_score(y_test,
                                            models[i].fit(X_train, 
                                                         y_train).predict(X_test)))
            r2_train.append(metrics.r2_score(y_train, 
                                             models[i].fit(X_train, 
                                                          y_train).predict(X_train)))
        results[i] = [np.mean(r2_train), np.mean(r2_test)]
    return pd.DataFrame(results)
models = {'OLS': linear_model.LinearRegression(),
         'Lasso': linear_model.Lasso(),
         'Ridge': linear_model.Ridge(),}
linmodels1_test = test(models, X_train, X_test, y_train, y_test)
print(linmodels1_test)

            OLS     Lasso     Ridge
0  8.095074e-01  0.408470  0.791106
1 -3.575315e+22  0.346233  0.412535


**<font color='teal'> Print the intercept value from the linear model. </font>**

In [95]:
lin_model.intercept_[0]

14971359374263.58

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [96]:

coef_table['abs_coef']= abs(coef_table.coefficent)
coef_table= coef_table.sort_values(by='abs_coef', ascending= False)
coef_table



variable    coefficent      abs_coef
7                 double  1.309203e+14  1.309203e+14
12         LongestRun_mi  6.692951e+13  6.692951e+13
1          vertical_drop  6.665541e+13  6.665541e+13
6                 triple -4.949525e+13  4.949525e+13
13     SkiableTerrain_ac -4.616809e+13  4.616809e+13
16             yearsOpen -3.817843e+13  3.817843e+13
9           total_chairs -3.562202e+13  3.562202e+13
11          TerrainParks -3.320994e+13  3.320994e+13
8                surface -3.000505e+13  3.000505e+13
15      daysOpenLastYear -2.516122e+13  2.516122e+13
4              fastQuads  2.378239e+13  2.378239e+13
5                   quad  2.162912e+13  2.162912e+13
3              fastSixes -1.906297e+13  1.906297e+13
20          state_Alaska -1.497136e+13  1.497136e+13
46    state_South Dakota -1.497136e+13  1.497136e+13
26        state_Illinois -1.497136e+13  1.497136e+13
42            state_Ohio -1.497136e+13  1.497136e+13
25           state_Idaho -1.497136e+13  1.497136e+13
35         state_Montana -1.497136e+13  1.497136e+13
54         state_Wyoming -1.497136e+13  1.497136e+13
27         state_Indiana -1.497136e+13  1.497136e+13
28            state_Iowa -1.497136e+13  1.497136e+13
43          state_Oregon -1.497136e+13  1.497136e+13
32        state_Michigan -1.497136e+13  1.497136e+13
33       state_Minnesota -1.497136e+13  1.497136e+13
40        state_New York -1.497136e+13  1.497136e+13
34        state_Missouri -1.497136e+13  1.497136e+13
31   state_Massachusetts -1.497136e+13  1.497136e+13
22      state_California -1.497136e+13  1.497136e+13
51      state_Washington -1.497136e+13  1.497136e+13
39      state_New Mexico -1.497136e+13  1.497136e+13
29           state_Maine -1.497136e+13  1.497136e+13
48            state_Utah -1.497136e+13  1.497136e+13
53       state_Wisconsin -1.497136e+13  1.497136e+13
24     state_Connecticut -1.497136e+13  1.497136e+13
21         state_Arizona -1.497136e+13  1.497136e+13
37   state_New Hampshire -1.497136e+13  1.497136e+13
52   state_West Virginia -1.497136e+13  1.497136e+13
50        state_Virginia -1.497136e+13  1.497136e+13
44    state_Pennsylvania -1.497136e+13  1.497136e+13
36          state_Nevada -1.497136e+13  1.497136e+13
41  state_North Carolina -1.497136e+13  1.497136e+13
49         state_Vermont -1.497136e+13  1.497136e+13
38      state_New Jersey -1.497136e+13  1.497136e+13
23        state_Colorado -1.497136e+13  1.497136e+13
45    state_Rhode Island -1.497136e+13  1.497136e+13
2                  trams -1.216644e+13  1.216644e+13
14        Snow Making_ac -1.010278e+13  1.010278e+13
18        NightSkiing_ac -9.731449e+12  9.731449e+12
19              clusters -6.081367e+12  6.081367e+12
17       averageSnowfall  4.247502e+12  4.247502e+12
10                  Runs -1.451049e+12  1.451049e+12
30        state_Maryland -1.269193e+12  1.269193e+12
0            summit_elev  3.934000e+10  3.934000e+10
47       state_Tennessee  0.000000e+00  0.000000e+00

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [97]:
#drop state
ski_data_trial2 = ski_data_trial.loc[:, ~ski_data_trial.columns.str.startswith('state')]

In [98]:

#rerun model excluding state variable
y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial2 , y, test_size=0.25, random_state=0)
lm = linear_model.LinearRegression()
lin_model2 = lm.fit(X_train,y_train)
y_pred2 = lm.predict(X_test)

#create Coefficent table
lin_model_coef2 = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef2[col_name] = lm.coef_[0][idx]   
coef_table2 =pd.DataFrame(lin_model_coef2.items(), columns=['variable', 'coefficent'])

#stats_new_model
lin_model2.intercept_[0]
lin_model2_Rsq = explained_variance_score(y_test, y_pred2)
lin_model2_me = mean_absolute_error(y_test, y_pred2)
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model2_Rsq, lin_model2_me))

linmodels2_test = test(models, X_train, X_test, y_train, y_test)

print('R2 for linear models that have state dummy variables')
print(linmodels1_test)

print('R2 for linear models that do not have state dummy variables')
print(linmodels2_test)


When state dummy variables are removed the Rsq of the linear model = 0.543151163811642 and the Mean Absolute Error = 8.071952560297293
R2 for linear models that have state dummy variables
            OLS     Lasso     Ridge
0  8.095074e-01  0.408470  0.791106
1 -3.575315e+22  0.346233  0.412535
R2 for linear models that do not have state dummy variables
        OLS     Lasso     Ridge
0  0.611043  0.374976  0.614314
1  0.542606  0.355843  0.551013


In [99]:
coef_table2['abs_coef']= abs(coef_table2.coefficent)
coef_table2= coef_table2.sort_values(by='abs_coef', ascending= False)
coef_table2

variable    coefficent      abs_coef
11       TerrainParks  2.371773e+15  2.371773e+15
9        total_chairs -1.629692e+15  1.629692e+15
16          yearsOpen -1.239544e+15  1.239544e+15
8             surface  1.165773e+15  1.165773e+15
5                quad -7.137372e+14  7.137372e+14
7              double -6.916187e+14  6.916187e+14
4           fastQuads -6.882206e+14  6.882206e+14
17    averageSnowfall -6.218395e+14  6.218395e+14
18     NightSkiing_ac -5.443262e+14  5.443262e+14
12      LongestRun_mi -3.493219e+14  3.493219e+14
10               Runs  3.278269e+14  3.278269e+14
6              triple  3.242365e+14  3.242365e+14
3           fastSixes -3.001582e+14  3.001582e+14
2               trams  2.761549e+14  2.761549e+14
19           clusters -2.272418e+14  2.272418e+14
1       vertical_drop  1.962869e+14  1.962869e+14
13  SkiableTerrain_ac  1.576476e+14  1.576476e+14
14     Snow Making_ac  4.144442e+13  4.144442e+13
15   daysOpenLastYear  8.620125e+11  8.620125e+11
0         summit_elev  6.307835e+00  6.307835e+00

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [100]:
ski_data_trial3 = ski_data_trial2.drop(['summit_elev'], axis =1)

#rerun model excluding state variable
y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial3 , y, test_size=0.25, random_state=0)
lm = linear_model.LinearRegression()
lin_model3 = lm.fit(X_train,y_train)
y_pred3 = lm.predict(X_test)

#create Coefficent table
lin_model_coef3 = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef3[col_name] = lm.coef_[0][idx]   
coef_table3 =pd.DataFrame(lin_model_coef3.items(), columns=['variable', 'coefficent'])

#stats_new_model
lin_model3.intercept_[0]
lin_model3_Rsq = explained_variance_score(y_test, y_pred3)
lin_model3_me = mean_absolute_error(y_test, y_pred3)
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model3_Rsq, lin_model3_me))

linmodels3_test = test(models, X_train, X_test, y_train, y_test)

print('R2 for linear models that have state dummy variables')
print(linmodels1_test)

print('R2 for linear models that do not have state dummy variables')
print(linmodels2_test)


print('R2 for linear models that do not have state dummy variables and summit elev')
print(linmodels3_test)

When state dummy variables are removed the Rsq of the linear model = 0.5352861270602318 and the Mean Absolute Error = 8.37660097515031
R2 for linear models that have state dummy variables
            OLS     Lasso     Ridge
0  8.095074e-01  0.408470  0.791106
1 -3.575315e+22  0.346233  0.412535
R2 for linear models that do not have state dummy variables
        OLS     Lasso     Ridge
0  0.611043  0.374976  0.614314
1  0.542606  0.355843  0.551013
R2 for linear models that do not have state dummy variables and summit elev
        OLS     Lasso     Ridge
0  0.606299  0.374975  0.604437
1  0.534110  0.355842  0.546046


## Identify the Final Model

In [86]:
print('The Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model_Rsq, lin_model_me))
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model2_Rsq, lin_model2_me))
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model3_Rsq, lin_model3_me))


The Rsq of the linear model = 0.3898894046548559 and the Mean Absolute Error = 623337519347.5455
When state dummy variables are removed the Rsq of the linear model = 0.543151163811642 and the Mean Absolute Error = 8.071952560297293
When state dummy variables are removed the Rsq of the linear model = 0.5352861270602318 and the Mean Absolute Error = 8.37660097515031


**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.3898894046548559 | 10.10875602049704  |-|
| Model 2. | 0.5402501392547869 | 8.129332457111145  |'state'|
| Model 3. | 0.5335568743843538 | 8.426849467905704 |'state','summit_elev','base_elev'|

Model Selection: Model3